tmt/steps/prepare/artifact/providers/repository_provider.py

```diff
@@ -1,108 +1,144 @@
+        # Cache for the list of RPMs discovered in the repository
+        self._rpm_list: Optional[list[dict[str, Any]]] = None
+        self._repository_installed = False
 
     def _parse_artifact_id(self, artifact_id: str) -> str:
-        # Validate the artifact_id by creating a RepositoryFileInfo instance
-        RepositoryFileInfo(_raw_artifact=artifact_id)
+        """
+        Validate the artifact identifier using the RepositoryFile class.
+        """
+        # The constructor of RepositoryFile handles URL validation
+        RepositoryFile(url=artifact_id)
         return artifact_id
 
-    def list_artifacts(self) -> Iterator[RepositoryFileInfo]:
-        yield RepositoryFileInfo(_raw_artifact=self.artifact_id)
+    def _fetch_rpms(self, guest: Guest, repo_filepath: Path) -> None:
+        """
+        Query the guest to find all packages available in the new repository.
+        """
+        # TODO: This method needs to be implemented to populate self._rpm_list with RPMs
+        # from the repository. Currently using an empty list as a temporary solution.
+        self._rpm_list = []
+
+    def list_artifacts(self) -> Iterator[RpmArtifactInfo]:
+        """
+        List all RPMs available from the repository.
+
+        Note: This requires the repository to be installed and queried first,
+        which is handled by the `download_artifacts` method.
+        """
+        if self._rpm_list is None:
+            raise GeneralError(
+                "RPM list not fetched. Call 'download_artifacts' to install "
+                "the repository and populate the package list."
+            )
+        for rpm in self._rpm_list:
+            yield RpmArtifactInfo(_raw_artifact=rpm)
 
     def _download_artifact(
-        self, artifact: RepositoryFileInfo, guest: Guest, destination: Path
+        self, artifact: RpmArtifactInfo, guest: Guest, destination: Path
     ) -> None:
-        url = artifact.location
-        filename = artifact.id
-        repo_destination = Path("/etc/yum.repos.d") / filename
-        sudo_prefix = "sudo " if not guest.facts.is_superuser else ""
-
-        self.logger.debug(f"Processing repository file from {url} to {repo_destination} on guest.")
+        """This provider only sets up the repo, it does not download RPMs."""
+        self.logger.debug(f"Skipping download of '{artifact.id}'.")
+
+    def download_artifacts(
+        self,
+        guest: Guest,
+        download_path: Path,
+        exclude_patterns: list[Pattern[str]],
+    ) -> list[Path]:
+        """
+        Download the .repo file to the guest, making the repository available.
+
+        This method overrides the default behavior to prevent downloading
+        individual RPMs. It installs the repository, lists the available
+        packages for discovery, and then returns.
+        """
+        if self._repository_installed:
+            self.logger.debug("Repository file already installed, skipping.")
+            return []
```

Comment:
Interesting. I would expect `download_artifacts` to be called only once, or at least only once per `guest`.